Docs/fixes from tfe202504 review (#888)
hashicorp/terraform-docs-common: website/docs/cloud-docs/vcs/azure-devops-services-pat.mdx

---
page_title: Set up the Azure DevOps Services VCS provider using personal access tokens
description: >-
  Learn how to use Azure DevOps Services with workspaces and private registry modules in HCP Terraform using personal access tokens.
---

# Set up the Azure DevOps Services VCS provider using personal access tokens

These instructions describe how to use a personal access token to connect HCP Terraform to Azure DevOps Services VCS, which is hosted at `dev.azure.com`. Connecting to Azure DevOps Services VCS lets you take advantage of HCP Terraform's VCS features. 

If you do not intent to use personal access tokens to connect to Azure DevOps Services, refer to [Set up the Azure DevOps Server VCS provider](/terraform/cloud-docs/vcs/azure-devops-server). For information about other supported VCS providers, refer to [Connect to VCS Providers](/terraform/cloud-docs/vcs).

## Requirements

Configuring a new VCS provider requires permission to [manage VCS settings](/terraform/cloud-docs/users-teams-organizations/permissions#manage-vcs-settings) for the organization.

> The account you use for connecting HCP Terraform must have **Project Collection Administrator** access enabled for any projects containing repositories of Terraform configurations. This permission is required to create webhooks. You cannot create custom access roles with lower levels of privilege because Microsoft does not allow delegation of this capability.

Reviewer: ```suggestion
To connect HCP Terraform to your repositories, your account needs **Project Collection Administrator** access for all projects containing Terraform configurations. The **Project Collection Administrator** permission is required to create webhooks. Microsoft doesn't allow creating custom roles with fewer permissions for this purpose.
```

[permissions-citation]: #intentionally-unused---keep-for-maintainers

## Personal access token maintenance

HCP Terraform uses personal access tokens to connect to Azure DevOps Services. This access method requires some additional configuration and ongoing maintenance:

> - Personal access tokens have a maximum lifetime of one year. When the token expires, HCP Terraform cannot to connect to Azure DevOps Services until the token is replaced. To avoid a gap in service, do one of the following actions before the token expires:

Reviewer: ```suggestion
- Personal access tokens have a maximum lifetime of one year. When the token expires, HCP Terraform cannot connect to Azure DevOps Services until the token is replaced. To avoid a gap in service, do one of the following actions before the token expires:
```

### Limit the scope of this VCS provider

1. Enable **Selected Projects** and use the text field that appears to search for and select projects to enable. All current and future workspaces for any selected projects can use repositories from this VCS Provider.
1. Click **Update VCS Provider** to save your selections.

1. Click the **Update VCS Provider** button to save your selections.

### Add an SSH key pair

HCP Terraform only uses SSH to clone Git submodules and uses HTTPS for all other Git operations.
Do not use your personal SSH key to connect HCP Terraform and Azure DevOps Services. You should generate a new key or use an existing key reserved for service access.

In the following steps, you must provide HCP Terraform with the private key. Although HCP Terraform does not display the text of the key to users after it is entered, HCP Terraform retains the key and uses for authenticating with Azure DevOps Services.

<Warning>

> Follow security best practices and your organization's policies for protecting important credentials when handing and storing the private key. Someone with access to the key can use it to push code to the repositories you use to manage your infrastructure.

Reviewer: ```suggestion
Follow security best practices and your organization's policies for protecting important credentials when handing and storing a private key. Someone with access to the key can use it to push code to the repositories you use to manage your infrastructure.
```

</Warning>

1. On a secure workstation, create an SSH keypair that HCP Terraform can use to connect to the Azure DevOps Services domain. The following example uses the `ssh-keygen` command to create a `service_terraform` file with the private key and a `service_terraform.pub` file with the public key:

  ```shell-session
  $ ssh-keygen -t rsa -m PEM -f "/Users/<NAME>/.ssh/service_terraform" -C "service_terraform_enterprise"
  ```

  <Warning>

>   The SSH key must have an empty passphrase. HCP Terraform cannot use SSH keys that require a passphrase.

Author: ```suggestion
  The passphrase for the SSH key must be empty. HCP Terraform cannot use SSH keys that require a passphrase.
```
